File: src/ppocrv6_min/det.py
# -*- coding: utf-8 -*-
"""DB text-line detector (PP-OCRv6 *_det ONNX).

Post-processing follows PaddleOCR ``ppocr/postprocess/db_postprocess.py``
exactly: threshold -> contours -> minAreaRect -> polygon-mean score ->
pyclipper ``unclip`` expansion.
"""
from __future__ import annotations

import os

import cv2
import numpy as np
import onnxruntime as ort
import pyclipper

# ImageNet stats (PaddleX inference.yml: NormalizeImage)
MEAN = np.array([0.485, 0.456, 0.406], dtype=np.float32)
STD = np.array([0.229, 0.224, 0.225], dtype=np.float32)


class DBPostProcess:
    """Contour -> scored unclipped quads, identical to PaddleOCR's DBPostProcess."""

    def __init__(self, thresh: float = 0.2, box_thresh: float = 0.4,
                 unclip_ratio: float = 1.4, max_candidates: int = 3000,
                 min_size: int = 3):
        self.thresh = thresh
        self.box_thresh = box_thresh
        self.unclip_ratio = unclip_ratio
        self.max_candidates = max_candidates
        self.min_size = min_size

    def unclip(self, box, ratio: float):
        box = np.asarray(box, dtype=np.float64)
        x = box[:, 0]
        y = box[:, 1]
        area = 0.5 * abs(np.dot(x, np.roll(y, -1)) - np.dot(y, np.roll(x, -1)))
        closed = np.vstack([box, box[:1]])
        length = float(np.linalg.norm(np.diff(closed, axis=1), axis=1).sum())
        distance = area * ratio / (length + 1e-9)
        offset = pyclipper.PyclipperOffset()
        offset.AddPath(box, pyclipper.JT_ROUND, pyclipper.ET_CLOSEDPOLYGON)
        return offset.Execute(distance)

    def get_mini_boxes(self, contour):
        bounding_box = cv2.minAreaRect(contour)
        points = sorted(list(cv2.boxPoints(bounding_box)), key=lambda x: x[0])
        i1, i2, i3, i4 = 0, 1, 2, 3
        if points[1][1] > points[0][1]:
            i1, i4 = 0, 1
        else:
            i1, i4 = 1, 0
        if points[3][1] > points[2][1]:
            i2, i3 = 2, 3
        else:
            i2, i3 = 3, 2
        box = [points[i1], points[i2], points[i3], points[i4]]
        return box, min(bounding_box[1])

    def box_score_fast(self, pred: np.ndarray, box) -> float:
        h, w = pred.shape
        box = np.array(box, dtype=np.float32).copy()
        xmin = int(np.clip(np.floor(box[:, 0].min()), 0, w - 1))
        xmax = int(np.clip(np.ceil(box[:, 0].max()), 0, w - 1))
        ymin = int(np.clip(np.floor(box[:, 1].min()), 0, h - 1))
        ymax = int(np.clip(np.ceil(box[:, 1].max()), 0, h - 1))
        mask = np.zeros((ymax - ymin + 1, xmax - xmin + 1), dtype=np.uint8)
        box[:, 0] -= xmin
        box[:, 1] -= ymin
        cv2.fillPoly(mask, box.reshape(1, -1, 2).astype("int32"), 1)
        return float(cv2.mean(pred[ymin:ymax + 1, xmin:xmax + 1], mask)[0])

    def __call__(self, pred: np.ndarray, dest_w: int, dest_h: int):
        """``pred``: (H, W) probability map at network output resolution."""
        mask = (pred > self.thresh).astype(np.uint8)
        contours, _ = cv2.findContours(mask * 255, cv2.RETR_LIST,
                                       cv2.CHAIN_APPROX_SIMPLE)
        h, w = pred.shape
        boxes, scores = [], []
        for contour in contours[: self.max_candidates]:
            points, sside = self.get_mini_boxes(contour)
            if sside < self.min_size:
                continue
            points = np.array(points)
            score = self.box_score_fast(pred, points)
            if self.box_thresh > score:
                continue
            box = self.unclip(points, self.unclip_ratio)
            if len(box) > 1:
                continue
            box = np.array(box).reshape(-1, 1, 2)
            box, sside = self.get_mini_boxes(box)
            if sside < self.min_size + 2:
                continue
            box = np.array(box, dtype="int32")
            box[:, 0] = np.clip(np.round(box[:, 0] / w * dest_w), 0, dest_w)
            box[:, 1] = np.clip(np.round(box[:, 1] / h * dest_h), 0, dest_h)
            boxes.append(box)
            scores.append(score)
        return boxes, scores


class DetModel:
    """DB text-line detector.

    Args:
        model_dir: directory holding ``inference.onnx`` (PaddleX det export).
        limit_side_len: long-side target for ``DetResizeForTest`` (default 960;
            the long side is never up-scaled beyond the original).
        thresh / box_thresh / unclip_ratio: DBPostProcess parameters
            (defaults match the PaddleX ``inference.yml`` of PP-OCRv6_tiny_det).
        providers: onnxruntime execution providers.
    """

    def __init__(self, model_dir: str, limit_side_len: int = 960,
                 thresh: float = 0.2, box_thresh: float = 0.4,
                 unclip_ratio: float = 1.4, providers=None):
        onnx_path = os.path.join(model_dir, "inference.onnx")
        if not os.path.isfile(onnx_path):
            raise FileNotFoundError(onnx_path)
        self.limit = limit_side_len
        self.post = DBPostProcess(thresh=thresh, box_thresh=box_thresh,
                                  unclip_ratio=unclip_ratio)
        self.sess = ort.InferenceSession(
            onnx_path, providers=providers or ["CPUExecutionProvider"])
        self.in_name = self.sess.get_inputs()[0].name

    def _resize(self, h: int, w: int):
        scale = min(self.limit / max(h, w), 1.0)
        nh = max(32, (int(h * scale) // 32) * 32)
        nw = max(32, (int(w * scale) // 32) * 32)
        return nh, nw

    def detect(self, img_bgr: np.ndarray):
        """Detect text lines.

        Args:
            img_bgr: HxWx3 uint8 BGR image.

        Returns:
            list of ``(points, score)`` where ``points`` is a (4, 2) float32
            quad (tl, tr, br, bl order is NOT guaranteed) in original image
            coordinates and ``score`` is the mean probability inside the box.
        """
        h, w = img_bgr.shape[:2]
        nh, nw = self._resize(h, w)
        im = cv2.resize(img_bgr, (nw, nh), interpolation=cv2.INTER_LINEAR)
        x = im.astype(np.float32)
        x = (x / 255.0 - MEAN) / STD
        x = x.transpose(2, 0, 1)[None]
        prob = self.sess.run(None, {self.in_name: x})[0][0, 0]  # (nh, nw)
        boxes, scores = self.post(prob, w, h)
        return [(b.astype(np.float32), float(s)) for b, s in zip(boxes, scores)]

File: src/det.js
// det.js — DB text-line detector (PP-OCRv6 *_det ONNX), Node port of ppocrv6_min.det.
//
// Post-processing follows PaddleOCR ppocr/postprocess/db_postprocess.py
// exactly: threshold -> contours -> minAreaRect -> polygon-mean score ->
// unclip expansion. The pyclipper offset is replaced by an analytic
// rectangle expansion (see DBPostProcess.unclip for why this is exact).

import fs from "node:fs";
import path from "node:path";
import * as ort from "onnxruntime-node";
import {
  resizeBilinear,
  findContours,
  minAreaRect,
  boxPoints,
  fillPolyMask,
  meanMasked,
} from "./cv.js";

// ImageNet stats (PaddleX inference.yml: NormalizeImage)
export const MEAN = [0.485, 0.456, 0.406];
export const STD = [0.229, 0.224, 0.225];

export class DBPostProcess {
  constructor({ thresh = 0.2, boxThresh = 0.4, unclipRatio = 1.4, maxCandidates = 3000, minSize = 3 } = {}) {
    this.thresh = thresh;
    this.boxThresh = boxThresh;
    this.unclipRatio = unclipRatio;
    this.maxCandidates = maxCandidates;
    this.minSize = minSize;
  }

  /**
   * Expand a closed polygon outward by `distance` (pyclipper
   * JT_ROUND + ET_CLOSEDPOLYGON equivalent).
   *
   * In the DB pipeline the input is always the output of minAreaRect, i.e. a
   * rectangle (possibly rotated). Offsetting a rectangle with round joins
   * yields a rounded rectangle whose minimum-area rectangle is exactly the
   * miter join of the edges moved out by `distance` — the rounded corners lie
   * strictly inside. pyclipper additionally rounds the input coordinates to
   * integers, which is reproduced here. The result is a single path or an
   * empty list, matching pyclipper.Execute.
   */
  unclip(box, ratio) {
    const pts = box.map((p) => [Math.round(p[0]), Math.round(p[1])]);
    // Shoelace area + perimeter on the float64 input (same as the Python code).
    let area = 0;
    let length = 0;
    for (let i = 0; i < box.length; i++) {
      const [x1, y1] = box[i];
      const [x2, y2] = box[(i + 1) % box.length];
      area += x1 * y2 - x2 * y1;
      length += Math.hypot(x2 - x1, y2 - y1);
    }
    area = 0.5 * Math.abs(area);
    const distance = (area * ratio) / (length + 1e-9);
    if (pts.length < 3) return [];
    if (distance === 0) return [pts]; // zero offset: pyclipper returns the input
    // Signed area determines orientation; outward normal = rotate(dir, 90°)
    // flipped so that it points away from the polygon centroid.
    const n = pts.length;
    let centroidX = 0;
    let centroidY = 0;
    for (const p of pts) {
      centroidX += p[0];
      centroidY += p[1];
    }
    centroidX /= n;
    centroidY /= n;
    const lines = [];
    for (let i = 0; i < n; i++) {
      const [ax, ay] = pts[i];
      const [bx, by] = pts[(i + 1) % n];
      const ex = bx - ax;
      const ey = by - ay;
      const len = Math.hypot(ex, ey);
      if (len === 0) continue;
      let nx = ey / len;
      let ny = -ex / len;
      const mx = (ax + bx) / 2 - centroidX;
      const my = (ay + by) / 2 - centroidY;
      if (nx * mx + ny * my < 0) {
        nx = -nx;
        ny = -ny;
      }
      lines.push({ px: ax + nx * distance, py: ay + ny * distance, dx: ex, dy: ey });
    }
    if (lines.length < 3) return [pts];
    const out = [];
    for (let i = 0; i < lines.length; i++) {
      const l1 = lines[(i - 1 + lines.length) % lines.length];
      const l2 = lines[i];
      out.push(intersectLines(l1, l2));
    }
    return [out];
  }

  /** cv2.minAreaRect + boxPoints + the PaddleOCR corner reordering. */
  getMiniBoxes(contour) {
    const rect = minAreaRect(contour);
    const points = boxPoints(rect).sort((a, b) => a[0] - b[0]);
    let i1; let i2; let i3; let i4;
    if (points[1][1] > points[0][1]) {
      i1 = 0; i4 = 1;
    } else {
      i1 = 1; i4 = 0;
    }
    if (points[3][1] > points[2][1]) {
      i2 = 2; i3 = 3;
    } else {
      i2 = 3; i3 = 2;
    }
    const box = [points[i1], points[i2], points[i3], points[i4]];
    return { box, sside: Math.min(rect.size[0], rect.size[1]) };
  }

  /** Mean probability inside `box` (float32 polygon) on the (h, w) prob map. */
  boxScoreFast(pred, w, h, box) {
    const pts = box.map((p) => [p[0], p[1]]);
    let minX = Infinity; let maxX = -Infinity; let minY = Infinity; let maxY = -Infinity;
    for (const p of pts) {
      if (p[0] < minX) minX = p[0];
      if (p[0] > maxX) maxX = p[0];
      if (p[1] < minY) minY = p[1];
      if (p[1] > maxY) maxY = p[1];
    }
    const xmin = clampInt(Math.floor(minX), 0, w - 1);
    const xmax = clampInt(Math.ceil(maxX), 0, w - 1);
    const ymin = clampInt(Math.floor(minY), 0, h - 1);
    const ymax = clampInt(Math.ceil(maxY), 0, h - 1);
    const winW = xmax - xmin + 1;
    const winH = ymax - ymin + 1;
    const local = pts.map((p) => [Math.trunc(p[0] - xmin), Math.trunc(p[1] - ymin)]);
    const mask = fillPolyMask(winW, winH, local);
    // pred window
    let sum = 0;
    let count = 0;
    for (let yy = 0; yy < winH; yy++) {
      for (let xx = 0; xx < winW; xx++) {
        if (mask[yy * winW + xx]) {
          sum += pred[(ymin + yy) * w + (xmin + xx)];
          count++;
        }
      }
    }
    return count === 0 ? 0 : sum / count;
  }

  /**
   * pred: Float32Array (H*W) probability map at network output resolution.
   * Returns { boxes: number[][][] (int quads), scores: number[] }.
   */
  run(pred, w, h, destW, destH) {
    const mask = new Uint8Array(w * h);
    for (let i = 0; i < w * h; i++) mask[i] = pred[i] > this.thresh ? 1 : 0;
    const contours = findContours(mask, w, h);
    const boxes = [];
    const scores = [];
    for (const contour of contours.slice(0, this.maxCandidates)) {
      let { box, sside } = this.getMiniBoxes(contour);
      if (sside < this.minSize) continue;
      const score = this.boxScoreFast(pred, w, h, box);
      if (this.boxThresh > score) continue;
      const expanded = this.unclip(box, this.unclipRatio);
      if (expanded.length > 1) continue;
      let box2;
      let sside2;
      if (expanded.length === 0) {
        box2 = [];
        sside2 = 0;
      } else {
        ({ box: box2, sside: sside2 } = this.getMiniBoxes(expanded[0]));
      }
      if (sside2 < this.minSize + 2) continue;
      // int32 truncation + rescale to the original image size, as in Python.
      const quad = box2.map((p) => [
        clampInt(Math.round((Math.trunc(p[0]) / w) * destW), 0, destW),
        clampInt(Math.round((Math.trunc(p[1]) / h) * destH), 0, destH),
      ]);
      boxes.push(quad);
      scores.push(score);
    }
    return { boxes, scores };
  }
}

function clampInt(v, lo, hi) {
  return Math.min(hi, Math.max(lo, v));
}

/** Intersection of two (point, direction) lines. */
function intersectLines(l1, l2) {
  const denom = l1.dx * l2.dy - l1.dy * l2.dx; // cross(d1, d2)
  if (denom === 0) {
    // Parallel edges (degenerate): keep the shifted endpoint.
    return [l2.px, l2.py];
  }
  const t =
    ((l2.px - l1.px) * l2.dy - (l2.py - l1.py) * l2.dx) / denom;
  return [l1.px + l1.dx * t, l1.py + l1.dy * t];
}

/** DB text-line detector (PP-OCRv6 det ONNX export). */
export class DetModel {
  /**
   * @param {string} modelDir directory holding inference.onnx
   * @param {object} opts { limitSideLen=960, thresh, boxThresh, unclipRatio, providers }
   */
  constructor(modelDir, { limitSideLen = 960, thresh = 0.2, boxThresh = 0.4, unclipRatio = 1.4, providers } = {}) {
    const onnxPath = path.join(modelDir, "inference.onnx");
    if (!fs.existsSync(onnxPath)) {
      throw new Error(`model not found: ${onnxPath}`);
    }
    this.limit = limitSideLen;
    this.post = new DBPostProcess({ thresh, boxThresh, unclipRatio });
    this.session = null; // created lazily (async)
    this.onnxPath = onnxPath;
    this.providers = providers;
  }

  /** Async initializer — await this before calling detect(). */
  async init() {
    this.session = await ort.InferenceSession.create(this.onnxPath, {
      executionProviders: this.providers || ["cpu"],
    });
    this.inName = this.session.inputNames[0];
    return this;
  }

  /** det resize: long side -> limitSideLen (never upscaled), rounded to %32. */
  targetSize(h, w) {
    const scale = Math.min(this.limit / Math.max(h, w), 1.0);
    const nh = Math.max(32, Math.floor(Math.trunc(h * scale) / 32) * 32);
    const nw = Math.max(32, Math.floor(Math.trunc(w * scale) / 32) * 32);
    return [nh, nw];
  }

  /**
   * Detect text lines in a BGR image object.
   * @returns {Promise<Array<{quad: number[][], score: number}>>} quads are
   *   (4,2) int boxes in original image coordinates (corner order not
   *   guaranteed), score is the mean probability inside the box.
   */
  async detect(imgBgr) {
    const { width: w, height: h, data } = imgBgr;
    const [nh, nw] = this.targetSize(h, w);
    const resized = resizeBilinear({ width: w, height: h, channels: 3, data }, nw, nh);
    // Normalize + NCHW, channel order BGR (as fed by the Python reference).
    const c = 3;
    const x = new Float32Array(c * nh * nw);
    const plane = nh * nw;
    const rd = resized.data;
    for (let i = 0; i < plane; i++) {
      for (let ch = 0; ch < c; ch++) {
        x[ch * plane + i] = (rd[i * c + ch] / 255.0 - MEAN[ch]) / STD[ch];
      }
    }
    const input = new ort.Tensor("float32", x, [1, c, nh, nw]);
    const out = await this.session.run({ [this.inName]: input });
    const t = out[this.session.outputNames[0]];
    // prob = out[0][0, 0] -> (nh, nw)
    const probData = t.data;
    const prob = probData.subarray(0, nh * nw); // [1,1,nh,nw] layout
    const { boxes, scores } = this.post.run(
      prob instanceof Float32Array ? prob : Float32Array.from(prob),
      nw, nh, w, h
    );
    return boxes.map((quad, i) => ({ quad, score: scores[i] }));
  }
}
